File: src/data/courses/dsa/two-pointers-sliding-window/two-pointers-opposite-direction/sort-colors/code/python/sort_colors.py
# ============================================================
# Sort Colors (Dutch National Flag)
# ============================================================

from typing import List

# ============================================================
# Solution 1: Counting Sort — Two Pass
# Time: O(N) | Space: O(1)
# ============================================================
class Solution1:
    def sortColors(self, nums: List[int]) -> None:
        count = [nums.count(0), nums.count(1), nums.count(2)]
        idx = 0
        for color in range(3):
            for _ in range(count[color]):
                nums[idx] = color
                idx += 1

# ============================================================
# Solution 2: Dutch National Flag — Single Pass (Optimal)
# Time: O(N) | Space: O(1)
# ============================================================
class Solution2:
    def sortColors(self, nums: List[int]) -> None:
        lo, mid, hi = 0, 0, len(nums) - 1

        while mid <= hi:
            if nums[mid] == 0:
                nums[lo], nums[mid] = nums[mid], nums[lo]
                lo += 1
                mid += 1
            elif nums[mid] == 1:
                mid += 1
            else:  # nums[mid] == 2
                nums[mid], nums[hi] = nums[hi], nums[mid]
                hi -= 1
                # Don't advance mid — swapped element is unknown

File: src/data/courses/dsa/two-pointers-sliding-window/two-pointers-opposite-direction/sort-colors/code/java/SortColors.java
// ============================================================
// Sort Colors (Dutch National Flag)
// ============================================================

// ============================================================
// Solution 1: Counting Sort — Two Pass
// Time: O(N) | Space: O(1)
// ============================================================
class Solution1 {
    public void sortColors(int[] nums) {
        int[] count = new int[3];
        for (int n : nums) count[n]++;

        int idx = 0;
        for (int c = 0; c < 3; c++)
            for (int i = 0; i < count[c]; i++)
                nums[idx++] = c;
    }
}

// ============================================================
// Solution 2: Dutch National Flag — Single Pass (Optimal)
// Time: O(N) | Space: O(1)
// ============================================================
class Solution2 {
    public void sortColors(int[] nums) {
        int lo = 0, mid = 0, hi = nums.length - 1;

        while (mid <= hi) {
            if (nums[mid] == 0) {
                int t = nums[lo]; nums[lo] = nums[mid]; nums[mid] = t;
                lo++; mid++;
            } else if (nums[mid] == 1) {
                mid++;
            } else {
                int t = nums[mid]; nums[mid] = nums[hi]; nums[hi] = t;
                hi--;
            }
        }
    }
}
